File: dotlox/Scanner.cs

namespace dotlox
{
    public class Scanner
    {
        String src;
        List<Token> tokens = new List<Token>();
        int start = 0;
        int current = 0;
        int line = 0;

        static Dictionary<String, TokenType> keywords;
        static Scanner()
        {
            keywords = new Dictionary<String, TokenType>();
            keywords.Add("add", TokenType.AND);
            keywords.Add("class", TokenType.CLASS);
            keywords.Add("else", TokenType.ELSE);
            keywords.Add("false", TokenType.FALSE);
            keywords.Add("for", TokenType.FOR);
            keywords.Add("fun", TokenType.FUN);
            keywords.Add("if", TokenType.IF);
            keywords.Add("nil", TokenType.NIL);
            keywords.Add("or", TokenType.OR);
            keywords.Add("print", TokenType.PRINT);
            keywords.Add("return", TokenType.RETURN);
            keywords.Add("super", TokenType.SUPER);
            keywords.Add("this", TokenType.THIS);
            keywords.Add("true", TokenType.TRUE);
            keywords.Add("var", TokenType.VAR);
            keywords.Add("while", TokenType.WHILE);
        }

        public Scanner(String src)
        {
            this.src = src;
        }

        public List<Token> scanTokens()
        {
            while (!isAtEnd())
            {
                start = current;
                scanToken();
            }
            addToken(TokenType.EOF);
            return tokens;
        }

        void scanToken()
        {
            char c = advance();
            Console.WriteLine($"c: {c}, curr: {current}/{src[current]}, start: {start}/{src[start]}");
            switch (c)
            {
                case '(': addToken(TokenType.LEFT_PAREN); break;
                case ')': addToken(TokenType.RIGHT_PAREN); break;
                case '{': addToken(TokenType.LEFT_BRACE); break;
                case '}': addToken(TokenType.RIGHT_BRACE); break;
                case ',': addToken(TokenType.COMMA); break;
                case '.': addToken(TokenType.DOT); break;
                case '-': addToken(TokenType.MINUS); break;
                case '+': addToken(TokenType.PLUS); break;
                case ';': addToken(TokenType.SEMICOLON); break;
                case '*': addToken(TokenType.STAR); break;
                case '!': addToken(match('=') ? TokenType.BANG_EQUAL : TokenType.EQUAL); break;
                case '=': addToken(match('=') ? TokenType.EQUAL_EQUAL : TokenType.EQUAL); break;
                case '<': addToken(match('=') ? TokenType.LESS_EQUAL : TokenType.LESS); break;
                case '>': addToken(match('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER); break;
                case '/':
                    {
                        if (match('/'))
                        {
                            while (peek() != '\n' && !isAtEnd()) advance();
                        }
                        else
                        {
                            addToken(TokenType.SLASH);
                        }
                        break;
                    }
                case ' ':
                case '\r':
                case '\t':
                    break;
                case '\n':
                    line++;
                    break;
                case '"': ParseString(); break;
                case var x when Char.IsDigit(x):
                    {
                        ParseNumber();
                        break;
                    };
                case var x when (Char.IsLetter(x) || x == '_'):
                    {
                        ParseIdentifier();
                        break;
                    };
                default: Program.error(line, $"Unexpected Char: '{c}'"); break;
            }
        }

        void addToken(TokenType type)
        {
            addToken(type, null);
        }

        void addToken(TokenType type, Object literal)
        {
            Console.WriteLine($"{src} == {start} -> {current}");
            String text = src.Substring(start, current);
            tokens.Add(new Token(type, text, literal, line));
        }

        bool match(char expected)
        {
            if (isAtEnd()) return false;
            if (src[current] != expected) return false;
            current++;
            return true;
        }

        char peek()
        {
            if (isAtEnd()) return '\0';
            return src[current];
        }

        char advance()
        {
            return src[current++];
        }


        char peekNext()
        {
            if (current + 1 >= src.Length) return '\0';
            return src[current + 1];
        }

        void ParseString()
        {
            var startLine = line;
            while (peek() != '"' && !isAtEnd())
            {
                if (peek() == '\n') line++;
                advance();
            }

            if (isAtEnd())
            {
                Program.error(line, $"Unterminated String (Started line: {startLine})");
                return;
            }
            advance(); // Grab the closing Quote
            String value = src.Substring(start + 1, current - 1);
            addToken(TokenType.STRING, value);
        }

        void ParseNumber()
        {
            while (Char.IsDigit(peek())) advance();
            if (peek() == '.' && Char.IsDigit(peekNext()))
            {
                advance();
                while (Char.IsDigit(peek())) advance();
            }
            addToken(TokenType.NUMBER, Double.Parse(src.Substring(start, current)));
        }

        void ParseIdentifier()
        {
            while (Char.IsLetterOrDigit(peek())) { advance(); }
            Console.WriteLine("New ident found!");
            String text = src.Substring(start, current);
            TokenType type = TokenType.IDENT;
            keywords.TryGetValue(text, out type); // Modifies 'type' inplace.
            addToken(type);
        }

        bool isAtEnd() { return current >= src.Length; }


    }
}

File: plox/plox/Scanner.py
from typing import Any
from plox.Token import Token, TokenType, Keywords


class Scanner:
    def __init__(self, src: str):
        self.src = src
        self.start: int = 0
        self.current: int = 0
        self.line: int = 0
        self.tokens: list[Token] = []

    def scanTokens(self) -> list[Token]:
        while not self.isAtEnd:
            self._scanToken()
            self.start = self.current
        self.add_token(TokenType.EOF)
        return self.tokens

    def _scanToken(self):
        c = self.advance()
        match c:
            case "(":
                self.add_token(TokenType.LEFT_PAREN)
            case "(":
                self.add_token(TokenType.LEFT_PAREN)
            case ")":
                self.add_token(TokenType.RIGHT_PAREN)
            case "{":
                self.add_token(TokenType.LEFT_BRACE)
            case "}":
                self.add_token(TokenType.RIGHT_BRACE)
            case ",":
                self.add_token(TokenType.COMMA)
            case ".":
                self.add_token(TokenType.DOT)
            case "-":
                self.add_token(TokenType.MINUS)
            case "+":
                self.add_token(TokenType.PLUS)
            case ";":
                self.add_token(TokenType.SEMICOLON)
            case "*":
                self.add_token(TokenType.STAR)
            case "/":
                self.add_token(TokenType.SLASH)
            case "!":
                self.add_token(
                    TokenType.BANG_EQUAL if self.match("=") else TokenType.BANG
                )
            case "=":
                self.add_token(
                    TokenType.EQUAL_EQUAL if self.match("=") else TokenType.EQUAL
                )
            case "<":
                self.add_token(
                    TokenType.LESS_EQUAL if self.match("=") else TokenType.LESS
                )
            case ">":
                self.add_token(
                    TokenType.GREATER_EQUAL if self.match("=") else TokenType.GREATER
                )
            # Whitespace/Comments
            case "#":
                while self.peek() != "\n" and not self.isAtEnd:
                    self.advance()
            case " " | "\r" | "\t":
                pass
            case "\n":
                self.line += 1
            case '"':
                self.parse_str()
            case digit if digit.isdigit():
                self.parse_number()
            case ident if ident.isalpha() or ident == "_":
                self.parse_ident()

            case _:
                raise ValueError(f"Could not match {c} to a Token!")

    def add_token(self, type_: TokenType, literal: Any = None):
        self.tokens.append(
            Token(type_, self.src[self.start : self.current], literal, self.line)
        )

    @property
    def isAtEnd(self):
        return self.current >= len(self.src)

    def match(self, c: str) -> bool:
        if self.isAtEnd:
            return False
        if self.src[self.current] != c:
            return False
        self.current = +1
        return True

    def peek(self) -> str:
        if self.isAtEnd:
            return "\0"
        return self.src[self.current]

    def peek_next(self) -> str:
        if self.current + 1 >= len(self.src):
            return "\0"
        return self.src[self.current + 1]

    def advance(self) -> str:
        idx = self.current
        self.current = idx + 1
        return self.src[idx]

    def parse_str(self):
        while self.peek() != '"' and not self.isAtEnd:
            if self.advance() == "\n":
                self.line += 1
        if self.isAtEnd:
            ValueError(f"{self.line}:: Unterminated String.")

        self.advance()  # Grab closing quote
        string = self.src[self.start + 1 : self.current - 1]
        self.add_token(TokenType.STRING, string)

    def parse_number(self):
        while self.peek().isdigit():
            self.advance()
        if self.peek() == "." and self.peek_next().isdigit():
            self.advance()
            while self.peek().isdigit():
                self.advance()
        self.add_token(TokenType.NUMBER, float(self.src[self.start : self.current]))

    def parse_ident(self):
        while self.peek().isalnum():
            self.advance()
        ident = self.src[self.start : self.current]
        type = Keywords.get(ident, TokenType.IDENT)
        self.add_token(type)
